### ISSUE TITLE
Add option to compact leveldb to repo fsck

### ISSUE THREAD

**Issue by kevina** [ipfs/go-ipfs#3927]
See #3763 for context.

**Comment by jefft0** [comment]
Will it be possible to run fsck to compact while the daemon is running (online) in another terminal? (In my steady-state use case, the daemon is running in one terminal to serve webcam files, and another terminal adds new files every hour, which can occasionally call fsck to compact since the online daemon doesn't seem to do it.)

**Comment by Kubuxu** [comment]
@jefft0 when daemon is running the leveldb is autocompacted (low priority background routine of leveldb), in theory.

As you have shown in #3763 adding a lot of files with no-copy without daemon running means problems with compaction.

**Comment by jefft0** [comment]
In my experience, the low priority compaction didn't happen and my add script failed. I need to remove the uncertainty of "maybe will compact." That's why I asked, can another script do fsck compact while the daemon is running online?

**Comment by Kubuxu** [comment]
That depends on how @kevina implements it. If it isn't much of an issue, I think it should be possible to do it online.

**Comment by kevina** [comment]
Related to the original problem but not this issue, I am concerned on why auto compaction was not triggered.  Is these really something that runs in the background and needs the process to run for a while in order to trigger, or is something about our usage causing it to not trigger, there are a bunch of parameters controlling compaction and perhaps they need to be adjusted for our use case.  Or perhaps the problem is fixed in the latest version of leveldb (we haven't upgraded in a while).  @whyrusleeping thoughts.

**Comment by kevina** [comment]
@whyrusleeping you seam to know a little more than I do, does the ipfs daemon some how trigger compaction?  Is it a background thread in leveldb that does the compaction and won't run if you only use leveldb for short periods of time?

**Comment by whyrusleeping** [comment]
@kevina its a background thread in leveldb that triggers the compaction normally

**Comment by kevina** [comment]
@whyrusleeping for the `fsck` option, I take it it should require the daemon be offline.  I think of you compact the full range it will get a write lock on the database so nothing will be able to happen anyway while the compaction is running.

**Comment by whyrusleeping** [comment]
@kevina Yeah, thats probably the right way to do it.

**Comment by whyrusleeping** [comment]
@kevina still planning on working on this?

**Comment by kevina** [comment]
Yeah.  I kind of forget about these little tasks.

**Comment by kevina** [comment]
I looked into it and implementing this will be non-trivial.  The first problem will be to get to the underlying datastore, this is possible, but only after calling `NewDatastore` as go-ds-levelds exports the DB field (by accident) but not the type returned:
```go
type datastore struct {
	DB *leveldb.DB
}
```
The second problem would be to discover the datastores that need compacting.

ref: https://github.com/ipfs/go-ds-leveldb/blob/4f27fe0c470cb3762d7a242bcf2f8d151e3da7dd/datastore.go
